### PR TITLE
[spec/builtin-printf] Implement printf %d \'c to print the numeric value of char c

### PR DESCRIPTION
**Pull request by asokoloski**
Closes: #435

### CODE REVIEW

File: osh/builtin_printf.py [hunk 1]
```diff
@@ -207,15 +207,21 @@ def __call__(self, arg_vec):
...
             except ValueError:
-              # This works around the fact that in the arg recycling case, you have no spid.
-              if word_spid == runtime.NO_SPID:
-                self.errfmt.Print("printf got invalid number %r for this substitution", s,
-                                  span_id=part.type.span_id)
+              if s[0] in '\'"':
+                try:
+                  d = ord(s[1])
+                except IndexError:
+                  d = 0
```
Comment (andychu) [inline osh/builtin_printf.py]: I pulled this and ran the tests, and it works!  Comments:

* How about adding test coverage for this `d = 0` case too?  When you just pass a single or double quote?
* The test confused me because you used `\*`.  Maybe choose `a=97` instead?

I'm probably going to change the exceptions to if statements because of the "mycpp" translation, but that is only a little related to this PR.

Comment (asokoloski) [reply osh/builtin_printf.py]: Sure, will do.  I actually started out with `a=97` but tried to be cute with `*=42` -- but clarity trumps cuteness :)

Comment (asokoloski) [reply osh/builtin_printf.py]: * 114e57d6d1109835da06ebceffad58a1065f7c78 -- switch to using "a" character
* c874e2badadf297213961e4905bda1853e78f8d6 -- add spec test for d = 0 case
* a3d47f87f7a10ead084dcc4b73b736a6cdccbead -- avoided adding an instance of exception handling in this MR, at least.

**Comment by andychu** [comment]
I merged this, thanks a lot!  Not many people have made a spec test change so I'm happy to see you got through it :)

I did find one problem:

https://github.com/oilshell/oil/commit/42458b8a965c63c48b843ea627a7e0c4ad0ab875

I also tightened up some corner cases.  Most shells are kind of sloppy about corner cases, but OSH is tring to be a little stricter. 

I made it stricter like `ash` in the empty string case, and stricter like `mksh` in the single quote case.

I debated whether to error out if there is more than one byte, but I didn't change that.

Oh another corner case is what to do in the case of a utf-8 encoded unicode char.  Let me at least test that -- I probably won't implement it for now.

**Comment by andychu** [comment]
OK, bash and zsh support unicode, but the other shells don't.

https://github.com/oilshell/oil/commit/78e089f37d39356b8f16ff7f8ba4301f289a2980

OSH should support Unicode in this case, but I don't think it's that important.  And as mentioned it depends on #366.  Because we have some pretty cheesy Python utf-8 code in `osh/string_ops.py` :)  That code helped us write good tests, but probably needs to be replaced.

Thanks for your help on this!

**Comment by asokoloski** [comment]
Glad to help!  Sorry about being a bit sloppy; thanks for tidying up after me.

**Comment by andychu** [comment]
No problem!  I'm still maintaining these issue labels if you're looking for something to hack on:

https://github.com/oilshell/oil/labels/good%20first%20issue

Or ask me on Zulip if you have any questions!
